plugin/pkg/admission/webhook/admission.go

```diff
@@ -0,0 +1,186 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package webhook checks a webhook for configured operation admission
+package webhook
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/apis/apiserver"
+	"k8s.io/apiserver/pkg/util/webhook"
+	"k8s.io/client-go/rest"
+
+	"k8s.io/kubernetes/pkg/api"
+	admissionv1alpha1 "k8s.io/kubernetes/pkg/apis/admission/v1alpha1"
+
+	// install the clientgo admissiony API for use with api registry
+	_ "k8s.io/kubernetes/pkg/apis/admission/install"
+)
+
+var (
+	groupVersions = []schema.GroupVersion{
+		admissionv1alpha1.SchemeGroupVersion,
+	}
+
+	ErrWebhookRejected = errors.New("rejected by a webhook admission controller")
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register("GenericAdmissionWebhook", func(configFile io.Reader) (admission.Interface, error) {
+		plugin, err := NewGenericAdmissionWebhook()
+		if err != nil {
+			return nil, err
+		}
+
+		return plugin, nil
+	})
+}
+
+// NewGenericAdmissionWebhook returns a generic admission webhook plugin.
+func NewGenericAdmissionWebhook() (admission.Interface, error) {
+	return &GenericAdmissionWebhook{
+		Handler: admission.NewHandler(
+			admission.Connect,
+			admission.Create,
+			admission.Delete,
+			admission.Update,
+		),
+	}, nil
+}
+
+type HookSource interface {
+	List() ([]apiserver.ExternalAdmissionHook, error)
+}
+
+type ServiceResolver interface {
+	ResolveEndpoint(namespace, name string) (*url.URL, error)
+}
+
+// GenericAdmissionWebhook is an implementation of admission.Interface.
+type GenericAdmissionWebhook struct {
+	*admission.Handler
+	hookSource      HookSource
+	serviceResolver ServiceResolver
+}
+
+// Admit makes an admission decision based on the request attributes.
+func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) (err error) {
+	hooks, err := a.hookSource.List()
+	if err != nil {
+		return fmt.Errorf("failed listing hooks: %v", err)
+	}
+
+	var errCount int64
+	var failCount int64
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(hooks))
+	for i := range hooks {
+		go func(hook *apiserver.ExternalAdmissionHook) {
+			defer wg.Done()
+			allowed, err := a.callHook(hook, attr)
+			if err != nil {
+				atomic.AddInt64(&errCount, 1)
+				glog.Warningf("Failed calling webhook %v: %v", hook.Name, err)
+			} else if !allowed {
+				atomic.AddInt64(&failCount, 1)
+			}
+		}(&hooks[i])
+	}
+	wg.Wait()
+
+	if failCount > 0 {
+		return ErrWebhookRejected
```

Comment:
Is this descriptive enough?  An API caller has no idea why it was rejected and the only possible way to figure that out is by accessing the API server logs, which is likely not an option.  Is there any good way to display the error(s) back to the user?  _(To me, it would be nice if all webhook responses with `Status.Result.Reason` were returned to the user that way they could possibly remedy the issue.)_